configs/all.js

```diff
@@ -9,6 +9,10 @@ function filterRules(rules, predicate) {
   return fromEntries(entries(rules).filter((entry) => predicate(entry[1])));
 }
 
+/**
+ * @param {object} rules - rules object mapping rule name to rule module
+ * @returns {Record<string, 2>}
```

Comment:
can these types be a bit better than `object` and `string`?

Reply:
Currently this function return type is inferred as one of two things depending on if there are types for `object.fromentries` installed:
- `any` if it's not
- `{ [x: string]: number; [x: number]: number; [x: symbol]: number; }` if it is

Using `Record<string, 2>` (which is the same as `{ [x: string]: 2 }`) is a marked improvement as it *always* has this type now.

Can we do better? Technically yes - this could be defined generically if we want.
I'd have to research how to do this with JSDoc types but it's possible to define this in TS with generics [playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXzD0SwHNk4BBAZwFEYYcYAeAFXhAA8MRVgr4qGGFlQkAfAAoYyCCCoAueACVwjYKwA08NAGtUOAO6oxASkUrCMdQAM4UTAHoAJAG8WAX2taATGIDcAFABhKiC8HBUMhjwALwERKTkINR0DDASLvAB8PAA5ABGODiyUKgAtAAODBVlqFAAtiIkuYouDgBU8AB0PfDtDu4a2Xn5yBgYeGUAFlBUZRgAnhUgLfBtnT1dfQNDOblgU+A6IMBlcACOyFgRZXgHpSQgtzBnIFDAeBALq+vdvf2DYa5UCIKBRSrVOb1KAYA4QnA1RbLKg-Dp-LYA3Z5UCCaSYchNMqzKikVD1HgYVEbf47ALuEyBAIRKKBBwOeAAPQA-EEQmEAG5QCDIBBxZkQDAAbVydkchWKb3KVQRtQaTVyAF1WeycpyeUA)
```ts
function configureAsError<T extends string>(rules: Record<T, unknown>): Record<`react/${T}`, 2> { ... }
```

Depends on how strict and correct you would like the types to be really!

Reply:
This would be the most correct JSDoc that would strictly type the resulting objects:

```js
/**
 * @template {string} Key
 * @template Value
 * @param {Record<Key, Value>} rules
 * @param {(value: Value) => boolean} predicate
 * @returns {Record<Key, Value>}
 */
function filterRules(rules, predicate) {
  return fromEntries(entries(rules).filter((entry) => predicate(entry[1])));
}

/**
 * @template {string} T
 * @param {Record<T, unknown>} rules - rules object mapping rule name to rule module
 * @returns {Record<`react/${T}`, 2>}
 */
function configureAsError(rules) {
  return fromEntries(Object.keys(rules).map((key) => [`react/${key}`, 2]));
}
```

However this causes lint errors.
The JSDoc parser used by the `valid-jsdoc` lint rule isn't permissive enough and crashes on the JSDoc comment if I try to do some better types.

```
/Users/bjz/git/eslint-plugin-react/configs/all.js
   8:1  error  JSDoc syntax error  valid-jsdoc
  19:1  error  JSDoc syntax error  valid-jsdoc
```

Reply:
If we're running tsc on the jsdoc comments, we don't need the valid-jsdoc rule anymore, i think?

Reply:
Note I didn't add tsc here - just the annotations.
Which was why I added ones that were as simple as possible cos they weren't for repo validation.

If we want to go down the path of using JSDoc types everywhere in this project that's a whole other topic :)